[DOC-166] Added rust-postgres smart driver (#20127)
yugabyte/yugabyte-db: docs/content/preview/reference/drivers/rust/rust-postgres-reference.md

# For yb-tokio-postgres
$ cargo add yb-tokio-postgres
```

## Fundamentals

Learn how to perform common tasks required for Rust application development using the YugabyteDB rust smart driver.

### Load balancing connection properties

The following connection properties need to be added to enable load balancing:

- `load_balance` - enable [cluster-aware load balancing](../../../../drivers-orms/smart-drivers/#cluster-aware-connection-load-balancing) by setting this property to true; disabled by default.
- `topology_keys` - provide comma-separated geo-location values to enable [topology-aware load balancing](../../../../drivers-orms/smart-drivers/#topology-aware-connection-load-balancing). Geo-locations can be provided as `cloud.region.zone`. Specify all zones in a region as `cloud.region.*`. To designate fallback locations for when the primary location is unreachable, specify a priority in the form `:n`, where `n` is the order of precedence. For example, `cloud1.datacenter1.rack1:1,cloud1.datacenter1.rack2:2`.

By default, the driver refreshes the list of nodes every 300 seconds (5 minutes). You can change this value by including the `yb_servers_refresh_interval` parameter.

> Following are other connection properties offered with the rust smart driver:

Reviewer: It should be underscore. The properties are `fallback_to_topology_keys_only` and `failed_host_reconnect_delay_secs`

To specify topology keys, you set the `topology_keys` property to comma-separated values in the connection string or dictionary, as per the following example:

```sh
let url: String = String::from( "postgresql://localhost:5434/yugabyte?user=yugabyte&password=yugabyte&load_balance=true&topology_keys=cloud1.datacenter1.rack2", );
let conn = yb_postgres::Client::connect(&connection_url,NoTls,)?;
```

## Try it out

This tutorial shows how to use the async yb-tokio-postgres client with YugabyteDB. It starts by creating a three-node cluster with a [replication factor](../../../../architecture/docdb-replication/replication/#replication-factor) of 3. This tutorial uses the [yugabyted](../../../configuration/yugabyted/) utility.

Next, you use a rust application to demonstrate the driver's load balancing features.

For an example to use the synchronous yb-postgres client, see [Connect an application](../../../../drivers-orms/rust/yb-rust-postgres).

### Create a local cluster

> Create a universe with a 3-node RF-3 cluster with some fictitious geo-locations assigned. The placement values used are just tokens and have nothing to do with actual AWS cloud regions and zones.

Reviewer 2: ```suggestion
Create a universe with a 3-node RF-3 cluster with some fictitious geo-locations assigned. Place two nodes in one location, and the third in a separate location. The placement values used are just tokens and have nothing to do with actual AWS cloud regions and zones.
```

```sh
cd <path-to-yugabytedb-installation>
```

To create a multi-zone cluster, do the following:

> 1. Start the first node by running the yugabyted start command, passing in the `--cloud_location` and `--fault_tolerance` flags to set the node location details, as follows:

Author: @HarshDaryani896 - I am adding --fault_tolerance=zone to the command , so 
./bin/yugabyted start --advertise_address=127.0.0.3 \
        --join=127.0.0.1 \
        --base_dir=$HOME/yugabyte-{{< yb-version version="preview" >}}/node3 \
        --cloud_location=aws.us-east-1.us-east-1b \
        --fault_tolerance=zone

    }

    async fn createconn(
       numconn: usize,
       url: String,
    ) -> Result<(Vec<Client>, Vec<JoinHandle<()>>), Error> {
       let mut connectionstored: Vec<JoinHandle<()>> = Vec::with_capacity(numconn);
       let mut clientstored: Vec<Client> = Vec::with_capacity(numconn);
       for _i in 0..numconn {
           let connectionresult = createconnection(url.clone()).await;
           match connectionresult {
              Err(error) => return Err(error),
              Ok((connection, client)) => {
                  clientstored.push(client);
                  connectionstored.push(connection);
              }
           }
       }
       return Ok((clientstored, connectionstored));
    }

    async fn createconnection(url: String) -> Result<(tokio::task::JoinHandle<()>, Client), Error> {
       let (client, connection) = yb_tokio_postgres::connect(&url, NoTls).await?;

       // The connection object performs the actual communication with the database,
       // so spawn it off to run on its own.
       let handle = tokio::spawn(async move {
           if let Err(e) = connection.await {
               eprintln!("connection error: {}", e);
           }

>        });

Reviewer: The opening brackets are in line 206.

>        });

Reviewer 2: I only see { bracket on 206, where is the ) opened?

>        });

Reviewer: It opens after `spawn` -> tokio::spawn`(`

   num_of_connections();

   println!("End of Example");
   Ok(())
}

async fn createconn(
   numconn: usize,
   url: String,
) -> Result<(Vec<Client>, Vec<JoinHandle<()>>), Error> {
   let mut connectionstored: Vec<JoinHandle<()>> = Vec::with_capacity(numconn);
   let mut clientstored: Vec<Client> = Vec::with_capacity(numconn);
   for _i in 0..numconn {
       let connectionresult = createconnection(url.clone()).await;
       match connectionresult {
           Err(error) => return Err(error),
           Ok((connection, client)) => {
               clientstored.push(client);
               connectionstored.push(connection);
           }
       }
   }
   return Ok((clientstored, connectionstored));
}

async fn createconnection(url: String) -> Result<(tokio::task::JoinHandle<()>, Client), Error> {
   let (client, connection) = yb_tokio_postgres::connect(&url, NoTls).await?;

   // The connection object performs the actual communication with the database,
   // so spawn it off to run on its own.
   let handle = tokio::spawn(async move {
       if let Err(e) = connection.await {
           eprintln!("connection error: {}", e);
       }

>    });

Reviewer: The opening brackets are in line 289.

   Ok((handle, client))
}

pub(crate) fn num_of_connections() {
   for i in 1..4 {
       let url = "http://127.0.0.".to_owned() + &i.to_string() + ":13000/rpcz";
       let response = isahc::get(url);
       if response.is_err() {
           println!("127.0.0.{} = {}",i, 0);
       } else {
           let body = response.unwrap().text().unwrap();
           let c = body.matches("client backend").count();
           println!("127.0.0.{} = {}", i, c);
       }
   }
}
```

In this case the first two nodes should have 15 connections each, and the third node should have zero connections.

### Clean up

After you're done experimenting, run the following command to destroy the local cluster:

```sh
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-2.20.0.1/node1
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-2.20.0.1/node2
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-2.20.0.1/node3
```

## Configure SSL/TLS

> The YugabyteDB rust-postgres smart driver support for SSL is the same as for the upstream driver.

Reviewer 2: ```suggestion
The YugabyteDB Rust smart driver support for SSL is the same as for the upstream driver.
```

> The following table describes the connection parameters required to connect using SSL.

Author: @ddhodge - I have listed the limitations under Configure SSL/TLS because its relevant to SSL modes.

| Parameter | Description | Default |
| :-------- | :---------- | :------ |
| sslmode | Controls usage of TLS. If set to:<br /> <ul><li> disable, TLS will not be used.</li><li> prefer, TLS will be used if available, but not used otherwise.</li><li> require, TLS will be forced to be used.</li></ul> | prefer |

The following is an example connection URL for connecting to a YugabyteDB cluster with SSL encryption enabled:

```sh
"postgresql://127.0.0.1:5434/yugabyte?user=yugabyte&password=yugabyte&load_balance=true&sslmode=require"
```

If you created a cluster on [YugabyteDB Managed](/preview/yugabyte-cloud/), use the cluster credentials and download the [SSL Root certificate](/preview/yugabyte-cloud/cloud-connect/connect-applications/).

The following is an example application for connecting to a YugabyteDB cluster with SSL enabled:

> Add `yb-postgres-openssl = "0.5.0-yb-1"` and `openssl = "0.10.61"` dependencies in the `Cargo.toml` file before executing the application.

Reviewer: `yb-postgres = "0.19.7-yb-1-beta"` also needs to be added to the dependencies, had missed this before.

```rust
use openssl::ssl::{SslConnector, SslMethod};
use yb_postgres_openssl::MakeTlsConnector;
use yb_postgres::{Client};

fn main()  {
   let mut builder = SslConnector::builder(SslMethod::tls()).expect("unable to create sslconnector builder");
   builder.set_ca_file("/path/to/root/certificate").expect("unable to load root certificate");
   let connector: MakeTlsConnector = MakeTlsConnector::new(builder.build());

   let mut connection = Client::connect( "host=? port=5433 dbname=yugabyte user=? password=? sslmode=require",
       connector,
       ).expect("failed to create tls ysql connection");

   let result = connection.query_one("select 1", &[]).expect("failed to execute select 1 ysql");
   let value: i32 = result.get(0);
   println!("result of query_one call: {}", value);
}
```

### Limitations

> Currently, PostgreSQL rust-postgres driver and YugabyteDB rust-postgres smart driver do not support verify-full and verify-ca ssl modes.

Reviewer 2: ```suggestion
Currently, the rust-postgres driver and YugabyteDB Rust smart driver do not support verify-full or verify-ca SSL modes.
```

After you're done experimenting, run the following command to destroy the local cluster:

```sh
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-{{< yb-version version="preview" >}}/node1
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-{{< yb-version version="preview" >}}/node2
./bin/yugabyted destroy --base_dir=$HOME/yugabyte-{{< yb-version version="preview" >}}/node3
```

## Configure SSL/TLS

The YugabyteDB Rust smart driver support for SSL is the same as for the upstream driver.

The following table describes the additional parameters the YugabyteDB Rust smart driver requires as part of the connection string when using SSL.

| rust-postgres parameter | Description | default |
| :---------------------- | :---------- | :------ |
| sslmode | SSL Mode | prefer |

> rust-postgres supports SSL modes in different ways depending on the driver version, as shown in the following table.

Reviewer 2: The rust-postgres driver supports the following SSL modes.

| SSL Mode | Description |
| :-------- | :---------- |
| disable | If set, TLS is not used. |
| prefer (default) | TLS is if available, but not otherwise. |
| require | If set, TLS is forced to be used. |

> 

Reviewer 2: Currently, the rust-postgres driver and YugabyteDB Rust smart driver do not support verify-full or verify-ca SSL modes.

and delete the limitations section below

| :---------------------- | :---------- | :------ |
| sslmode | SSL Mode | prefer |

The rust-postgres driver supports the following SSL modes.

| SSL Mode | Description |
| :-------- | :---------- |
| disable | If set, TLS is not used. |

> | prefer (default) | TLS is if available, but not otherwise. |

Reviewer 2: ```suggestion
| prefer (default) | Use TLS if available, but not otherwise. |
```
